pkg/openstack/instances.go

```diff
@@ -677,26 +728,86 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
+func getAddressesByName(compute *gophercloud.ServiceClient, name types.NodeName, networkingOpts NetworkingOpts, network *gophercloud.ServiceClient) ([]v1.NodeAddress, error) {
+	srv, err := getServerByName(compute, name)
 	if err != nil {
 		return nil, err
 	}
 
-	interfaces, err := getAttachedInterfacesByID(client, srv.ID)
+	interfaces, err := getAttachedInterfacesByID(compute, srv.ID, network)
 	if err != nil {
 		return nil, err
 	}
 
 	return nodeAddresses(&srv.Server, interfaces, networkingOpts)
 }
 
+// getSubInterfaces returns the sub-interfaces attached via trunks to the specified node interfaces
+func getSubInterfaces(interfaces []attachinterfaces.Interface, network *gophercloud.ServiceClient) ([]attachinterfaces.Interface, error) {
+	klog.V(5).Infof("Node has %d interfaces '%v'", len(interfaces), interfaces)
+
+	var subports []attachinterfaces.Interface
+	for _, iface := range interfaces {
+		// Check if trunk ports are attached
+		listOpts := trunks.ListOpts{
+			PortID: iface.PortID,
+		}
+		allPages, err := trunks.List(network, listOpts).AllPages()
+		if err != nil {
+			klog.Errorf("Failed to list trunks: %v", err)
+			return subports, nil
+		}
+		allTrunks, err := trunks.ExtractTrunks(allPages)
+		if err != nil {
+			klog.Errorf("Failed to extract trunks: %v", err)
+			return subports, err
+		}
+		// Get subports attached to the trunks
+		for _, trunk := range allTrunks {
+			if len(trunk.Subports) == 0 {
+				continue
+			}
+			klog.V(5).Infof("Subports for trunk %s: %v", trunk.ID, trunk.Subports)
+			// Arrange subports as for directly attached ports
+			for _, sport := range trunk.Subports {
+				p, err := ports.Get(network, sport.PortID).Extract()
+				if err != nil {
+					klog.Errorf("Failed to get port info for subport %s: %v", sport.PortID, err)
+					return subports, err
+				}
+				n, err := networks.Get(network, p.NetworkID).Extract()
+				if err != nil {
+					klog.Errorf("Failed to get network info for subport %s: %v", sport.PortID, err)
+					return subports, err
+				}
+				var sface = attachinterfaces.Interface{
+					PortState: "ACTIVE",
+					FixedIPs:  []attachinterfaces.FixedIP{},
+					PortID:    p.ID,
+					NetID:     n.Name,
+					MACAddr:   p.MACAddress,
+				}
+				for _, ip := range p.FixedIPs {
+					var ip2 = attachinterfaces.FixedIP{
+						SubnetID:  ip.SubnetID,
+						IPAddress: ip.IPAddress,
+					}
+					sface.FixedIPs = append(sface.FixedIPs, ip2)
+				}
+				subports = append(subports, sface)
+			}
+		}
+	}
+	klog.V(5).Infof("Node has %d sub-interfaces '%v'", len(subports), subports)
+	return subports, nil
+}
+
 // getAttachedInterfacesByID returns the node interfaces of the specified instance.
-func getAttachedInterfacesByID(client *gophercloud.ServiceClient, serviceID string) ([]attachinterfaces.Interface, error) {
+func getAttachedInterfacesByID(compute *gophercloud.ServiceClient, serviceID string, network *gophercloud.ServiceClient) ([]attachinterfaces.Interface, error) {
```

Comment:
nit: Again, I'd group the service clients.

Reply:
comment taken,